dbusmock/templates/accounts_service.py

```diff
@@ -0,0 +1,392 @@
+
+__author__ = 'Marco Trevisan'
+__email__ = 'marco.trevisan@canonical.com'
+__copyright__ = '(c) 2021 Canonical Ltd.'
+__license__ = 'LGPL 3+'
+
+import dbus
+
+from dbusmock import MOCK_IFACE, mockobject
+
+BUS_NAME = 'org.freedesktop.Accounts'
+MAIN_OBJ = '/org/freedesktop/Accounts'
+MAIN_IFACE = 'org.freedesktop.Accounts'
+USER_IFACE = MAIN_IFACE + '.User'
+SYSTEM_BUS = True
+
+
+def get_user_path(uid):
+    return '/org/freedesktop/Accounts/User{}'.format(uid)
+
+
+def load(mock, parameters=None):
+    parameters = parameters if parameters else {}
+    mock.mock_users = {}
+    mock.cached_users = []
+    mock.automatic_login_users = []
+    mock.users_auto_uids = 2000
+
+    mock.AddProperties(MAIN_IFACE, mock.GetAll(MAIN_IFACE))
+
+    for uid, name in parameters.get('users', {}).items():
+        mock.AddUser(uid, name)
+
+
+def emit_properties_changed(mock, interface=MAIN_IFACE, properties=None):
+    if properties is None:
+        properties = mock.GetAll(interface)
+    elif isinstance(properties, str):
+        properties = [properties]
+
+    if isinstance(properties, (list, set)):
+        properties = {p: mock.Get(interface, p) for p in properties}
+    elif isinstance(properties, dict):
+        pass
+
+    mock.EmitSignal(dbus.PROPERTIES_IFACE, 'PropertiesChanged', 'sa{sv}as', (
+        interface, properties, []))
+
+
+@dbus.service.method(MOCK_IFACE, in_signature='xsa{sv}', out_signature='o')
+def AddUser(self, uid, username, overrides=None):
+    '''Add user via uid and username and optionally overriding properties
+
+    Returns the new object path.
+   '''
+    path = get_user_path(uid)
+    default_props = {
+        'Uid': dbus.UInt64(uid),
+        'UserName': username,
+        'RealName': username[0].upper() + username[1:] + ' Fake',
+        'AccountType': dbus.Int32(1),
+        'AutomaticLogin': False,
+        'BackgroundFile': '',
+        'Email': '{}@python-dbusmock.org'.format(username),
+        'FormatsLocale': 'C',
+        'HomeDirectory': '/tmp/dbusmock-home/{}'.format(username),
+        'IconFile': '',
+        'InputSources': dbus.Array([], signature='a{ss}'),
+        'Language': 'C',
+        'LocalAccount': True,
+        'Location': '',
+        'Locked': False,
+        'LoginFrequency': dbus.UInt64(0),
+        'LoginHistory': dbus.Array([], signature='(xxa{sv})'),
+        'LoginTime': dbus.Int64(0),
+        'PasswordHint': 'Remember it, come on!',
+        'PasswordMode': 0,
+        'Session': 'mock-session',
+        'SessionType': 'wayland',
+        'Shell': '/bin/zsh',
+        'SystemAccount': False,
+        'XHasMessages': False,
+        'XKeyboardLayouts': dbus.Array([], signature='s'),
+        'XSession': 'mock-xsession',
+    }
+    default_props.update(overrides if overrides else {})
+    self.AddObject(path, USER_IFACE, default_props, [
+        ('SetRealName', 's', '', SetRealName),
+        ('SetEmail', 's', '', SetEmail),
+        ('SetLanguage', 's', '', SetLanguage),
```

Comment:
These trivial setters could just be written inline here. There's a lot of them, they all look the same, and with full-blown functions they create a lot of redundancy and overhead (like duplicating the signature).

Reply:
Ok, I tend to prefer to use real methods as they have some better static checks imho, so we can avoid duplications via https://github.com/martinpitt/python-dbusmock/pull/72 (and jonas utility method to just add a method via its handle), avoiding all the duplication I guess, no?

Reply:
I'll review #72, thanks. I still don't get why you want/need to *both* enumerate them here *and* define them as proper functions. Please pick just one. I don't understand what "better static checks" a trivial function like
```
@dbus.service.method(USER_IFACE, in_signature='s')
def SetEmail(self, email):
    set_user_property(self, 'Email', email)
```
does -- that does exactly the same (i.e. hardly any) static checks than

    ('SetEmail', 's', '', 'self.set_user_property("Email", args[0])')

and avoids the boilerplate. But anyway, if you prefer the full-blown functions I don't veto that, but then please remove them from the AddObject() call.

Reply:
> and avoids the boilerplate. But anyway, if you prefer the full-blown functions I don't veto that, but then please remove them from the AddObject() call.

Ohhh... So, I thought that this was required in order to make them work, for some reason I think they didn't work when I initially did this, thus the reason of listing them (and eventually of #72).